File: workers/typescript/harness/src/profiles.ts
import type { WorkerOptions } from '@temporalio/worker';

export const WORKER_PROFILE_ENV_VAR = 'OMES_WORKER_PROFILE';
export const RESOURCE_BASED_DEFAULT_PROFILE = 'resource-based-default';

export type WorkerProfile = Readonly<Partial<WorkerOptions>>;

const profiles = new Map<string, WorkerProfile>();

function registerWorkerProfile(name: string, profile: WorkerProfile): void {
  profiles.set(name, profile);
}

export function lookupWorkerProfile(name: string): Partial<WorkerOptions> {
  const profile = profiles.get(name);
  if (profile === undefined) {
    throw new Error(`Unknown worker profile "${name}"`);
  }
  return { ...profile };
}

registerWorkerProfile(RESOURCE_BASED_DEFAULT_PROFILE, {
  tuner: {
    tunerOptions: {
      targetMemoryUsage: 0.8,
      targetCpuUsage: 0.8,
    },
  },
});

File: workers/dotnet/Harness/Profiles.cs
using Temporalio.Worker;
using Temporalio.Worker.Tuning;
using WorkerProfile = Temporalio.Worker.TemporalWorkerOptions;

namespace Temporalio.Omes.Projects.Harness;

internal static class WorkerProfiles
{
    public const string EnvVarName = "OMES_WORKER_PROFILE";
    public const string ResourceBasedDefaultProfile = "resource-based-default";
    public const string ThroughputStressBaselineProfile = "throughput-stress-baseline";

    private static readonly IReadOnlyDictionary<string, WorkerProfile> Profiles =
        new Dictionary<string, WorkerProfile>
        {
            [ResourceBasedDefaultProfile] = new TemporalWorkerOptions
            {
                Tuner = WorkerTuner.CreateResourceBased(
                    targetMemoryUsage: 0.8,
                    targetCpuUsage: 0.8),
            },
            [ThroughputStressBaselineProfile] = new TemporalWorkerOptions
            {
                MaxCachedWorkflows = 50,
                MaxConcurrentWorkflowTasks = 8,
                MaxConcurrentActivities = 32,
                MaxConcurrentLocalActivities = 32,
                MaxConcurrentWorkflowTaskPolls = 2,
                MaxConcurrentActivityTaskPolls = 4,
            },
        };

    public static WorkerProfile LookupWorkerProfile(string name)
    {
        if (!Profiles.TryGetValue(name, out var profile))
        {
            throw new ArgumentException($"Unknown worker profile \"{name}\"");
        }

        return (WorkerProfile)profile.Clone();
    }

}
